MGMT-130 competency models (#14)
packbackbooks/engineering-docs: careers/technical-engineer.md

> # Technical Engineer Path

Reviewer: I would suggest that Senior+ Engineers certainly do take on sprint goal related work. They're not always DRI'ing or leading a project, it's possible that they're participating in a project that someone else is DRI'ing and as a result, are contributing to sprint goals.

This is how things currently work on my team.

> The technical engineer focuses on individual contributorship and technical leadership. Their contributions level up the quality, effectiveness and efficiency of the entire organization. Whereas the [software engineer path](software-engineer.md) focused on mastering the fundamentals of software development using a defined set of criteria, the senior technical path focuses more on projects/efforts that have strategic value for the company while driving technical design and up-leveling their team. As engineers progress down this path, examples of success will be less pre-defined and more dependent on the contextual needs of the business.

Author: Love it.

## Senior Engineer

The senior engineer displays a mastery of the core competencies of software engineering. Strong technical knowledge and high productivity are table stakes at this level. Growth opportunities focus on developing impact, communication, and collaboration skills in preparation for the staff engineer role.

> **Technical Knowledge:** They have deep understanding of their technical specialty, and a working of knowledge of other domains. They take responsibility for the overall health and quality of their domain. They have responsibilities beyond writing code (managing audits, running on-call rotations, owning a backlog, etc.) to help them manage their domain within the organization. Feedback on their code usually revolves around subjective decisions, rather than objective issues. They proactively stay abreast of industry best practices and trends, and have acquired at least one major skill outside of core coding such as monitoring, documentation, integration testing, visual design, performance optimization.

Reviewer 2: Minor addition to make the list clearer:

```suggestion
**Technical Knowledge:** They have deep understanding of their technical specialty, and a working of knowledge of other domains. They take responsibility for the overall health and quality of their domain. They have responsibilities beyond writing code (managing audits, running on-call rotations, owning a backlog, etc.) to help them manage their domain within the organization. Feedback on their code usually revolves around subjective decisions, rather than objective issues. They proactively stay abreast of industry best practices and trends, and have acquired at least one major skill outside of core coding such as monitoring, documentation, integration testing, visual design, and performance optimization.
```

> **Technical Knowledge:** They have deep understanding of their technical specialty, and a working of knowledge of other domains. They take responsibility for the overall health and quality of their domain. They have responsibilities beyond writing code (managing audits, running on-call rotations, owning a backlog, etc.) to help them manage their domain within the organization. Feedback on their code usually revolves around subjective decisions, rather than objective issues. They proactively stay abreast of industry best practices and trends, and have acquired at least one major skill outside of core coding such as monitoring, documentation, integration testing, visual design, and performance optimization.

Author: EN3s have done it historically in practice. That said, usually it's our more senior EN3s who are trending towards promotion already

**Productivity:** A senior engineer regularly delivers a high volume of complex, critical work while also balancing other responsibilities or initiatives. The Senior Engineer requires very little oversight beyond high-level direction. They hold a depth of knowledge about the larger software ecosystem that enables them to debug quickly and effectively.

> **Impact:** Beyond just delivering work on our roadmap, a senior engineer helps implement innovations across our codebase, product, and department. They recognize and act upon opportunities that others miss. They seek to build objective and well-defended cases for decisions, and develop a reputation among the leadership team that their requests are objective, self-critical and have due-diligence put in place to understand all viable alternatives.

Reviewer 2: I think this reads a little more clearly and gives a more active role on the engineer:

```suggestion
**Impact:** Beyond just delivering work on our roadmap, a senior engineer helps implement innovations across our codebase, product, and department. They recognize and act upon opportunities that others miss. They seek to build objective and well-defended cases for decisions, and develop a reputation among the leadership team that their requests are objective, self-critical, and have the due-diligence to understand all viable alternatives.
```

**Collaboration:** They ensure that they do not become a single point of failure for their domain by mentoring and training others. When relevant, they work effectively with others outside the organization (whether in other departments or with vendors). They are capable of effectively running projects of any size independently.

## Staff Engineer

**Technical Direction:** Technology cannot speak for itself and requires effective advocates on its behalf. Staff engineers successfully advance technology by being pragmatic, deliberate, and focused more on the long-term trend of progress than viewing each individual decision as a make-or-break crisis. Staff engineers essentially act like part-time product manager for technology. They are able to get everyone focused on a goal and lay out the steps to get there.

Concrete examples could include:

- Lead or delegate extra-large projects across teams
- Oversee cross-team initiatives (a11y, code quality, testing, reliability, etc.)
- Solve problems that span multiple groups

> **Engineering Perspective:** Staff-plus engineers are the folks who will often get unexpectedly pulled into the room where this sort of decision is happening. This gives them the opportunity to inject the engineering context and perspective into a decision while it's still possible to change the outcome. These brief moments of input on critical decisions are unduly impactful and will allow you to inject an engineering perspective where it would otherwise be missed.

Reviewer 3: `this sort of decision` is unclear - what sort of decision?

Concrete examples could include:

> - Your perspective is sought out in critical decisions

Reviewer 3: Having an example of "Engineering Perspective" be "your perspective was given" doesn't feel like much of an example. Is there maybe an example of a critical decision where a Staff Engineer's perspective would be desired / required?

The other three examples here are kind of in the same boat - this feels more like restating the paragraph above rather than concrete examples.

> - Actively inserting yourself into these critical moments

Reviewer 2: Shifting to the 2nd-person seems a little weird to me since every other section is in 3rd-person; can that be updated?

> - Advocate for effective solutions and strategies

Reviewer 2: The other examples have active `-ing` verbs:

```suggestion
- Advocating for effective solutions and strategies
```

- Bringing perspective from other teams or departments

**Mentorship:**

A Staff Engineer is far more likely to change your company's long-term trajectory by growing other engineers than through personal heroics. Their effort here leads to a well-trained team with multiple experts in each domain.

> Effective mentorship is not a checkbox on the career ladder that can be satisfied by simply sharing advice and experience. Staff engineers build an ongoing relationship, understand the context of their mentee, and [sponsor](https://larahogan.me/blog/what-sponsorship-looks-like/) them.

Reviewer 3: ```suggestion
Effective mentorship is not a checkbox on the career ladder that can be satisfied by simply sharing advice and experience. Staff engineers build an ongoing relationship, understand the context of their mentee(s), and [sponsor](https://larahogan.me/blog/what-sponsorship-looks-like/) them.
```

> Effective mentorship is not a checkbox on the career ladder that can be satisfied by simply sharing advice and experience. Staff engineers build an ongoing relationship, understands the context of their mentee, and [sponsors](https://larahogan.me/blog/what-sponsorship-looks-like/) them.

Reviewer 2: Minor grammatical changes ("staff engineers" is plural):

```suggestion
Effective mentorship is not a checkbox on the career ladder that can be satisfied by simply sharing advice and experience. Staff engineers build an ongoing relationship, understand the context of their mentee, and [sponsor](https://larahogan.me/blog/what-sponsorship-looks-like/) them.
```

Concrete examples could include:

- Owning onboarding and continuing education initiatives

> - No bus-factor-one concerns / build redundancy in systems and knowledge

Reviewer 2: ```suggestion
- No bus-factor-one concerns / building redundancy in systems and knowledge
```

- Providing critical feedback on architectural design documents

> - Sponsorship

Reviewer 3: Again, sponsorship is mentioned above. An example of sponsorship isn't sponsorship. Maybe we could say something like "Encouraging other engineers to take a specific online training or certification course"

Author: Lol. Good catch

**Exploration:**

> Our teams are already optimized to incrementally improve. The role of the staff engineer is to be *transformative*. A staff engineer does more than improve their part of the codebase; they solve ambiguous, important problems that the company's current systems are ill-shaped to address. They tackle impossible technical problems (high ambiguity, no in-house expertise, poorly defined). They identify and reduce existential risks. They look beyond their team and project to address organization-level problems (2-3 year growth, org health). They set the technical direction for a department.

Reviewer 3: I _love_ this explanation. Giving me goosebumps.

Concrete examples could include:

- Reduce infrastructure costs by an order of magnitude
- Identify a multi-region strategy that takes six months instead of three years
- Addressing the sudden realization that your primary database only has three months of remaining disk space, and you can't upgrade to a larger size
- Altering our CI/CD strategy to improve DORA metrics by 50% across the board
- Identifying a software vendor that reduces churn and increases revenue by 10%

**Be Glue:**

Perhaps most difficult to define, Staff Engineers do [glue work](https://noidea.dog/glue). They see the important things that need done and make sure they happen (whether doing it themselves or finding the right person or process for the job).

Concrete examples could include:

> - decreased onboarding time
> - got the team to agree on coding standards
> - reviewed designs and architecture decisions
> - made us build the right thing
> - made our customers happier
> - raised the organization's profile through public speaking/writing
> - simplified security and compliance concerns

Reviewer 2: The prior examples are all in the present tense, but these are in the past; how does this sound?

```suggestion
- decreasing onboarding time
- getting the team to agree on coding standards
- reviewing designs and architecture decisions
- making us build the right thing
- making our customers happier
- raising the organization's profile through public speaking/writing
- simplifying security and compliance concerns
```

> ## Principal Engineer

Reviewer 2: I can't comment directly on the lines since they're outside the diff, but I've got some suggestions for this section.

1. 1st paragraph: add a comma after "QA":
> "(including but not limited to... engineering, QA, and data analysis)
2. 2nd paragraph: replace "is" with "are"
> "The problems that the Principal Engineer ~~is~~ are solving..."
3. 4th paragraph - this is very nit-picky, but I think "both" is a little confusing because one interpretation can be that there are two core projects, whereas it's actually referring to "Packback" and "side and experimental work". Maybe this is clearer?
> "...they are a prolific contributor to both Packback's core projects and their own external work."
4. 4th paragraph: I think adding some commas to this list would make it clearer:
> "When presented with a complex problem, process, or existing system, they..."
5. 4th paragrap: In the context of the words after this phrase, I think "leave" doesn't quite fit since it describes how the person should feel after the experience; IMO we should either swap it out with another word or tweak the subsequent phrase.

Swapping out "leave":
> "...working with this person should ~~leave~~ have team members feeling like..."

Tweaking the subsequent phrase:
> "...working with this person should leave team members feeling like they ~~are going to leave with~~ gained something significantly better than they came into.

6. 5th paragraph: I think repositioning the comma can make this sentence clearer.
> "They create architecture that shapes large parts of our business and ship complex projects, including many systems or major pieces of infrastructure."

7. 5th paragraph: The last sentence feels a little passive, what do you think of this?
> "They are the origin of multiple obvious and large technical contributions."

8. 6th paragraph: I think using semicolons instead of commas can make this sentence clearer.
> "... capable of setting short, medium, and long term strategic direction for parts of the technology stack; identifying areas of critical need based on future growth; and developing roadmaps to attack those problems."

# Technical Engineer Path

The technical engineer focuses on individual contributorship and technical leadership. Their contributions level up the quality, effectiveness and efficiency of the entire organization. Whereas the [software engineer path](software-engineer.md) focused on mastering the fundamentals of software development using a defined set of criteria, the senior technical path focuses more on projects/efforts that have strategic value for the company while driving technical design and up-leveling their team. Their impact is no longer additive, it is multiplicative. As engineers progress down this path, examples of success will be less pre-defined and more dependent on the contextual needs of the business.

## Senior Engineer

> The senior engineer displays a mastery of the core competencies of software engineering. Strong technical knowledge and high productivity are table stakes at this level. Growth opportunities focus on developing impact, communication, and collaboration skills in preparation for the staff engineer role.

Reviewer: I can't think of a better way to phrase this, so we should probably keep as is. Yet, having the growth opportunities oriented around getting promoted to the next role seems incorrect. 

1. There are growth opportunities in the current role that aren't about going to the next (Senior Engineer is a completely acceptable stopping point career wise, not everyone wants to go to Staff).
2. If growth is just about getting to the next level, we've not properly articulated the desired outcomes of growth.

**Technical Knowledge:** They have deep understanding of their technical specialty, and a working knowledge of other domains. They take responsibility for the overall health and quality of their domain. They have responsibilities beyond writing code (managing audits, running on-call rotations, owning a backlog, etc.) to help them manage their domain within the organization. Feedback on their code usually revolves around subjective decisions, rather than objective issues. They proactively stay abreast of industry best practices and trends, and have acquired at least one major skill outside of core coding such as monitoring, documentation, integration testing, visual design, and performance optimization.

**Productivity:** A senior engineer regularly delivers a high volume of complex, critical work while also balancing other responsibilities or initiatives. The Senior Engineer requires very little oversight beyond high-level direction. They hold a depth of knowledge about the larger software ecosystem that enables them to debug quickly and effectively.

> **Impact:** Beyond just delivering work on our roadmap, a senior engineer helps implement innovations across our codebase, product, and department. They recognize and act upon opportunities that others miss. They seek to build objective and well-defended cases for decisions, and develop a reputation among the leadership team that their requests are objective, self-critical, and have the due-diligence to understand all viable alternatives.

Reviewer: This is pretty good, just need to drop "They seek to build objective and well-defended....". The "outcome" here is either they do this or they don't, I'd expect a Senior to do this.

**Technical Direction:** Technology cannot speak for itself and requires effective advocates on its behalf. Staff engineers successfully advance technology by being pragmatic, deliberate, and focused more on the long-term trend of progress than viewing each individual decision as a make-or-break crisis. Staff engineers essentially act like part-time product managers for technology. They are able to get everyone focused on a goal and lay out the steps to get there.

Concrete examples could include:

- Lead or delegate extra-large projects across teams
- Oversee cross-team initiatives (a11y, code quality, testing, reliability, etc.)
- Solve problems that span multiple groups

**Engineering Perspective:** Staff-plus engineers are the folks who will often get unexpectedly pulled into the room where this sort of decision is happening. This gives them the opportunity to inject the engineering context and perspective into a decision while it's still possible to change the outcome. These brief moments of input on critical decisions are unduly impactful and will allow you to inject an engineering perspective where it would otherwise be missed.

Concrete examples could include:

> - Having your perspective sought out in critical decisions

Reviewer: What's an example of a critical decision? Who's seeking out Staff Eng. perspective? VP? CTO? CEO?

> - Actively inserting yourself into these critical moments

Reviewer: What does this mean in practice? Once we define critical moments, I'd like to understand what our expectations are around Staff Eng. "inserting [themselves]" into those critical moments.

Concrete examples could include:

- Reduce infrastructure costs by an order of magnitude
- Identify a multi-region strategy that takes six months instead of three years
- Addressing the sudden realization that your primary database only has three months of remaining disk space, and you can't upgrade to a larger size
- Altering our CI/CD strategy to improve DORA metrics by 50% across the board
- Identifying a software vendor that reduces churn and increases revenue by 10%

**Be Glue:**

Perhaps most difficult to define, Staff Engineers do [glue work](https://noidea.dog/glue). They see the important things that need done and make sure they happen (whether doing it themselves or finding the right person or process for the job).

Concrete examples could include:

- decreasing onboarding time
- getting the team to agree on coding standards
- reviewing designs and architecture decisions
- making us build the right thing

> - making our customers happier

Reviewer: These are not concrete, so I would make them so or drop them from the list of concrete examples.

- raising the organization's profile through public speaking/writing
- simplifying security and compliance concerns

## Principal Engineer

> This person acts as the "chief architect" for our business. They have significant strategic vision and can take a high-level 3-5 year plan for growth at a business level and translate that into a strategic technology roadmap. They are deeply technically savvy, and their primary job is focusing on the architectural and technological needs to grow the business over the longer-term horizon. The principal engineer is able to plan and execute on tasks across all of Packback's engineering domains (including but not limited to: frontend dev, backend dev, site-reliability-engineering, QA, and data analysis)

Author: Similar to the managing engineer path, the Principal Engineer title needs a revisit, but I want to do that another time. The changes made to this section were mostly to ensure consistent voice.

**Communication:** The Principal Engineer has excellent abilities to influence without requiring reporting authority to do so. They effectively facilitate cross-team work and are influential far beyond their individual group. They are capable of driving groups of disparate interests to decisions, and clearly communicating and seeing those decisions through to impact. The Principal Engineer is capable of setting short, medium, and long term strategic direction for parts of the technology stack; identifying areas of critical need based on future growth; and developing roadmaps to attack those problems.

**Responsibilities:**

- Defining strategic technical initiatives
- Solving technical problems of the highest scope and complexity for the entire organization

> - Actively seeking out and prioritize our toughest technical challenges with a goal of creating significant improvement

Reviewer: Instead of "with a goal of creating significant improvement", can we just say "and delivering significant improvement"

**Responsibilities:**

- Define strategic technical initiatives
- Solve technical problems of the highest scope and complexity for the entire organization
- Actively seek out and prioritize our toughest technical challenges with a goal of creating significant improvement
- Define code quality standards and metrics
- Provide mentorship to Senior and Staff Engineers to help them grow in their technical responsibilities
- Ensure the safety and reliability of Packback's technology

> - Manage projects across multiple years

Reviewer 2: I'd personally prefer if these used active `-ing` verbs for consistency with other sections, but it's not a big deal either way.
